### PR TITLE
Golangci lint

### PR DESCRIPTION
**Pull request by bege13mot**
#730

### CODE REVIEW

**Comment by rhatdan** [comment]
@bege13mot You need sign all of your PRs
Also needs a rebase.

**Comment by bege13mot** [comment]
> @bege13mot You need sign all of your PRs
> Also needs a rebase.

Thank you, I've signed all commits, but unfortunately, I'm not sure about rebasing. Could you please help a little bit with it?

**Comment by vrothberg** [comment]
You need to rebase on the most recent commit on master. There are many ways to do that. I usually keep the upstream projects as a remote:
```bash
$ git remote add upstream git@github.com:containers/image.git
$ git remote update
# now can rebase our branch on top of upstream/master
$ git pull --rebase upstream master
```

**Comment by bege13mot** [comment]
Thank you

**Comment by bege13mot** [comment]
Unfortunately, still can't pass the test:
`
cmd/skopeo/layers.go:145:28: not enough arguments in call to dest.PutManifest
2035	have (context.Context, []byte)
2036	want (context.Context, []byte, *digest.Digest)
2037cmd/skopeo/layers.go:149:20: not enough arguments in call to dest.Commit
2038	have (context.Context)
2039	want (context.Context, types.UnparsedImage)
`

**Comment by vrothberg** [comment]
I restarted Travis. We had to update Skopeo to unblock the CI here.

**Comment by vrothberg** [comment]
Commit https://github.com/containers/image/pull/731/commits/eae61fa0aa3bc43529505c8a6663e66021d29f6b must be dropped from this PR.

File: openshift/openshift-copies.go [hunk 1]
```diff
@@ -210,13 +209,13 @@ func (config *directClientConfig) ClientConfig() (*restConfig, error) {
 		if err != nil {
 			return nil, err
 		}
-		mergo.MergeWithOverwrite(clientConfig, userAuthPartialConfig)
+		_ = mergo.MergeWithOverwrite(clientConfig, userAuthPartialConfig)
```
Comment (vrothberg) [inline openshift/openshift-copies.go]: I think we should check these errors. I'm not familiar with the `openshift` transport but it doesn't seem right to silently ignore.

File: openshift/openshift-copies.go [hunk 2]
```diff
@@ -355,7 +346,9 @@ func validateClusterInfo(clusterName string, clusterInfo clientcmdCluster) []err
 	}
 	if len(clusterInfo.CertificateAuthority) != 0 {
 		clientCertCA, err := os.Open(clusterInfo.CertificateAuthority)
-		defer clientCertCA.Close()
+		defer func() {
+			_ = clientCertCA.Close()
```
Comment (vrothberg) [inline openshift/openshift-copies.go]: That's something we could log using `logrus.Debugf("error closing client cert CA: %v", err)`

File: openshift/openshift-copies.go [hunk 3]
```diff
@@ -394,17 +387,17 @@ func validateAuthInfo(authInfoName string, authInfo clientcmdAuthInfo) []error {
 
 		if len(authInfo.ClientCertificate) != 0 {
 			clientCertFile, err := os.Open(authInfo.ClientCertificate)
-			defer clientCertFile.Close()
 			if err != nil {
 				validationErrors = append(validationErrors, errors.Errorf("unable to read client-cert %v for %v due to %v", authInfo.ClientCertificate, authInfoName, err))
 			}
+			defer clientCertFile.Close()
```
Comment (vrothberg) [inline openshift/openshift-copies.go]: Why is this moved? If golangci-lint doesn't complain, then it doesn't belong in this PR.

File: openshift/openshift-copies.go [hunk 4]
```diff
@@ -394,17 +387,17 @@ func validateAuthInfo(authInfoName string, authInfo clientcmdAuthInfo) []error {
...
 		}
 		if len(authInfo.ClientKey) != 0 {
 			clientKeyFile, err := os.Open(authInfo.ClientKey)
-			defer clientKeyFile.Close()
 			if err != nil {
 				validationErrors = append(validationErrors, errors.Errorf("unable to read client-key %v for %v due to %v", authInfo.ClientKey, authInfoName, err))
 			}
+			defer clientKeyFile.Close()
```
Comment (vrothberg) [inline openshift/openshift-copies.go]: Why is this moved? If golangci-lint doesn't complain, then it doesn't belong in this PR.

File: openshift/openshift.go [hunk 5]
```diff
@@ -490,7 +490,7 @@ sigExists:
 			Type:       imageSignatureTypeAtomic,
 			Content:    newSig,
 		}
-		body, err := json.Marshal(sig)
+		body, _ := json.Marshal(sig)
```
Comment (vrothberg) [inline openshift/openshift.go]: We should check this error.

File: pkg/blobinfocache/internal/prioritize/prioritize_test.go [hunk 6]
```diff
@@ -11,12 +11,12 @@ import (
 )
 
 const (
-	digestUnknown             = digest.Digest("sha256:1111111111111111111111111111111111111111111111111111111111111111")
-	digestUncompressed        = digest.Digest("sha256:2222222222222222222222222222222222222222222222222222222222222222")
-	digestCompressedA         = digest.Digest("sha256:3333333333333333333333333333333333333333333333333333333333333333")
-	digestCompressedB         = digest.Digest("sha256:4444444444444444444444444444444444444444444444444444444444444444")
-	digestCompressedUnrelated = digest.Digest("sha256:5555555555555555555555555555555555555555555555555555555555555555")
-	digestCompressedPrimary   = digest.Digest("sha256:6666666666666666666666666666666666666666666666666666666666666666")
+	// digestUnknown             = digest.Digest("sha256:1111111111111111111111111111111111111111111111111111111111111111")
```
Comment (vrothberg) [inline pkg/blobinfocache/internal/prioritize/prioritize_test.go]: Why is this commented out?

File: pkg/blobinfocache/internal/prioritize/prioritize_test.go [hunk 7]
```diff
@@ -11,12 +11,12 @@ import (
...
+	digestUncompressed = digest.Digest("sha256:2222222222222222222222222222222222222222222222222222222222222222")
+	digestCompressedA  = digest.Digest("sha256:3333333333333333333333333333333333333333333333333333333333333333")
+	digestCompressedB  = digest.Digest("sha256:4444444444444444444444444444444444444444444444444444444444444444")
+	// digestCompressedUnrelated = digest.Digest("sha256:5555555555555555555555555555555555555555555555555555555555555555")
```
Comment (vrothberg) [inline pkg/blobinfocache/internal/prioritize/prioritize_test.go]: Why is this commented out?

File: pkg/blobinfocache/internal/test/test.go [hunk 8]
```diff
@@ -17,7 +17,7 @@ const (
 	digestCompressedA         = digest.Digest("sha256:3333333333333333333333333333333333333333333333333333333333333333")
 	digestCompressedB         = digest.Digest("sha256:4444444444444444444444444444444444444444444444444444444444444444")
 	digestCompressedUnrelated = digest.Digest("sha256:5555555555555555555555555555555555555555555555555555555555555555")
-	digestCompressedPrimary   = digest.Digest("sha256:6666666666666666666666666666666666666666666666666666666666666666")
+	// digestCompressedPrimary   = digest.Digest("sha256:6666666666666666666666666666666666666666666666666666666666666666")
```
Comment (vrothberg) [inline pkg/blobinfocache/internal/test/test.go]: Why is this commented out?

File: signature/policy_reference_match_test.go [hunk 9]
```diff
@@ -57,7 +57,7 @@ func TestParseImageAndDockerReference(t *testing.T) {
 type refImageMock struct{ reference.Named }
 
 func (ref refImageMock) Reference() types.ImageReference {
-	return refImageReferenceMock{ref.Named}
+	return refImageReferenceMock(ref)
```
Comment (vrothberg) [inline signature/policy_reference_match_test.go]: I don't see golangci complaining about it. Why is that needed?

File: signature/signature.go [hunk 10]
```diff
@@ -112,8 +112,9 @@ var _ json.Unmarshaler = (*untrustedSignature)(nil)
 func (s *untrustedSignature) UnmarshalJSON(data []byte) error {
 	err := s.strictUnmarshalJSON(data)
 	if err != nil {
-		if _, ok := err.(jsonFormatError); ok {
+		if serror, ok := err.(jsonFormatError); ok {
```
Comment (vrothberg) [inline signature/signature.go]: I don't see golangci complaining about it. Why is it needed?

File: storage/storage_reference.go [hunk 11]
```diff
@@ -92,7 +92,7 @@ func imageMatchesSystemContext(store storage.Store, img *storage.Image, manifest
 		manifestType = manifest.GuessMIMEType(manifestBytes)
 	}
 	// Load the image's configuration blob.
-	m, err := manifest.FromBlob(manifestBytes, manifestType)
+	m, _ := manifest.FromBlob(manifestBytes, manifestType)
```
Comment (vrothberg) [inline storage/storage_reference.go]: We should check the error.

File: storage/storage_test.go [hunk 12]
```diff
@@ -248,6 +244,7 @@ func makeLayer(t *testing.T, compression archive.Compression) (ddigest.Digest, i
 	for i := 1024; i < 2048; i++ {
 		buf[i] = 0
 	}
+	errs := make(chan error, 1)
```
Comment (vrothberg) [inline storage/storage_test.go]: Why is that needed? golangci doesn't report it.

File: storage/storage_test.go [hunk 13]
```diff
@@ -513,7 +515,7 @@ func TestWriteRead(t *testing.T) {
 			if err != nil {
 				t.Fatalf("Error decompressing layer %q from %q", layerInfo.Digest, ref.StringWithinTransport())
 			}
-			n, err := io.Copy(&buf, decompressed)
+			n, _ := io.Copy(&buf, decompressed)
```
Comment (vrothberg) [inline storage/storage_test.go]: We should check the error.

File: signature/policy_reference_match_test.go (hunk 9 above)
Comment (bege13mot) [reply signature/policy_reference_match_test.go]: S1016: should convert ref (type refImageMock) to refImageReferenceMock instead of using struct literal

File: signature/signature.go (hunk 10 above)
Comment (bege13mot) [reply signature/signature.go]: S1020: when ok is true, err can't be nil

**Comment by bege13mot** [comment]
Hi @vrothberg ,
Thank you very much for your review.

File: openshift/openshift-copies.go [hunk 14]
```diff
@@ -272,13 +281,14 @@ func getUserIdentificationPartialConfig(configAuthInfo clientcmdAuthInfo) (*rest
...
-		len(config.BearerToken) > 0
+// Unused
+// // canIdentifyUser is a modified copy of k8s.io/kubernetes/pkg/client/unversioned/clientcmd.canIdentifyUser
+// func canIdentifyUser(config restConfig) bool {
+// 	return len(config.Username) > 0 ||
+// 		(len(config.CertFile) > 0 || len(config.CertData) > 0) ||
+// 		len(config.BearerToken) > 0
 
-}
+// }
```
Comment (TomSweeneyRedHat) [inline openshift/openshift-copies.go]: I'd prefer removing unused code unless you think there's a very real chance we'll need it again in the not too distant future.

**Comment by TomSweeneyRedHat** [comment]
CI tests are not at all happy.

File: storage/storage_test.go [hunk 15]
```diff
@@ -30,14 +30,12 @@ import (
...
 	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/require"
 )
 
 var (
-	_imgd      types.ImageDestination = &storageImageDestination{}
-	_imgs      types.ImageSource      = &storageImageSource{}
-	_ref       types.ImageReference   = &storageReference{}
-	_transport types.ImageTransport   = &storageTransport{}
```
Comment (mtrmac) [inline storage/storage_test.go]: These things are actually tests that the various structs implement the necessary interfaces. We should find a way to preserve that.

File: storage/storage_test.go [hunk 16]
```diff
@@ -30,14 +30,12 @@ import (
...
-	topwd                             = ""
+	// golangci-lint these variables are unused
```
Comment (mtrmac) [inline storage/storage_test.go]: I have no idea what this comment is trying to say. The variable _is_ used.

File: storage/storage_test.go [hunk 17]
```diff
@@ -248,6 +246,8 @@ func makeLayer(t *testing.T, compression archive.Compression) (ddigest.Digest, i
 	for i := 1024; i < 2048; i++ {
 		buf[i] = 0
 	}
+	// SA2002: the goroutine calls T.Fatalf, which must be called in the same goroutine as the test
+	errs := make(chan error, 1)
```
Comment (mtrmac) [inline storage/storage_test.go]: Don’t comment what the code _was_, say what it does _now_.

Did you mean something like “errs exists to return an error from the goroutine, because the `t.Fatalf` can’t be called by any goroutines newly created by the test”?

In that case I’d instead prefer the goroutine itself to start with a “Warning: don’t call `t.Fatalf` from this goroutine, it must be called from the primary goroutine”, or something like that — that’s what _everyone_ must know. The purpose of `errs` is less important, and follows naturally.

The name `errs` suggests that there could be _multiple_ values returned; that’s not what the intends to implement AFAICS.

File: storage/storage_test.go [hunk 18]
```diff
@@ -264,20 +264,24 @@ func makeLayer(t *testing.T, compression archive.Compression) (ddigest.Digest, i
...
-			t.Fatalf("Error writing tar header: %v", err)
+			errs <- fmt.Errorf("Error writing tar header: %v", err)
 		}
 		if n != len(buf) {
-			t.Fatalf("Short write writing tar header: %d < %d", n, len(buf))
+			errs <- fmt.Errorf("Short write writing tar header: %d < %d", n, len(buf))
 		}
 		err = twriter.Flush()
 		if err != nil {
 			t.Fatalf("Error flushing output to tar archive: %v", err)
```
Comment (mtrmac) [inline storage/storage_test.go]: If `Fatalf` really is forbidden, this one must be handled as well.

File: storage/storage_test.go [hunk 19]
```diff
@@ -264,20 +264,24 @@ func makeLayer(t *testing.T, compression archive.Compression) (ddigest.Digest, i
 			Typeflag:   tar.TypeReg,
 		})
 		if err != nil {
-			t.Fatalf("Error writing tar header: %v", err)
+			errs <- fmt.Errorf("Error writing tar header: %v", err)
```
Comment (mtrmac) [inline storage/storage_test.go]: This is not sufficient, the previous code (intended to) terminate the goroutine. Now it will just send something to `errs` and continue.


Really, rather than repeat the error handling all over the place (and risk forgetting), I’d much rather split into _two_ separate (named, top-level) functions:
- One to be the main function of the goroutine, dealing _only_ with the error reporting
- One to actually do the work, and just return an ordinary `error` return value, not caring about goroutines at all. 

Compare `copy.diffIDComputationGoroutine` and `copy.computeDiffID`.

File: storage/storage_test.go [hunk 20]
```diff
@@ -264,20 +264,24 @@ func makeLayer(t *testing.T, compression archive.Compression) (ddigest.Digest, i
...
 		if n != len(buf) {
-			t.Fatalf("Short write writing tar header: %d < %d", n, len(buf))
+			errs <- fmt.Errorf("Short write writing tar header: %d < %d", n, len(buf))
 		}
 		err = twriter.Flush()
 		if err != nil {
 			t.Fatalf("Error flushing output to tar archive: %v", err)
 		}
 	}()
+	err = <-errs
```
Comment (mtrmac) [inline storage/storage_test.go]: I can’t see how this doesn’t block forever if there is no error; `errs` is never closed, and nothing is sent into it.

(See above, a good reason to dedicate a separate function to dealing with the channel properly.)

File: types/types.go [hunk 21]
```diff
@@ -448,7 +448,7 @@ const (
 // OptionalBoolFalse.  The function is meant to avoid boilerplate code of users.
 func NewOptionalBool(b bool) OptionalBool {
 	o := OptionalBoolFalse
-	if b == true {
+	if b {
```
Comment (mtrmac) [inline types/types.go]: (I’m not thrilled about this one, but, oh well, I can see how the linter does not like the previous version.)

File: Makefile [hunk 22]
```diff
@@ -105,4 +105,4 @@ else
 	git fetch -q "https://github.com/containers/image.git" "refs/heads/master"
 	upstream="$$(git rev-parse --verify FETCH_HEAD)" ; \
 		git-validation -q -run DCO,short-subject,dangling-whitespace -range $$upstream..HEAD
-endif
+endif
```
Comment (mtrmac) [inline Makefile]: Please undo this.

File: openshift/openshift-copies.go [hunk 23]
```diff
@@ -394,14 +399,20 @@ func validateAuthInfo(authInfoName string, authInfo clientcmdAuthInfo) []error {
 
 		if len(authInfo.ClientCertificate) != 0 {
 			clientCertFile, err := os.Open(authInfo.ClientCertificate)
-			defer clientCertFile.Close()
+			defer func() {
+				err = clientCertFile.Close()
+				logrus.Debugf("error closing client cert CA: %v", err)
```
Comment (mtrmac) [inline openshift/openshift-copies.go]: The two copy&pasted errors here and below are incorrect.

**Comment by mtrmac** [comment]
> Hi, it looks like an issue with containers/skopeo
> 
> ```
> /tmp/tmp.hP4WQEvbic/src/github.com/containers/skopeo/hack/make/validate-lint: line 14: golint: command not found
> Makefile:175: recipe for target 'validate-local' failed
> ```

At a first glance, this looks like an issue in how _this_ PR changes the environment for `make test-skopeo` to me.

File: copy/copy.go [hunk 24]
```diff
@@ -762,7 +762,7 @@ func (ic *imageCopier) copyLayers(ctx context.Context) error {
 		defer progressCleanup()
 
 		for i, srcLayer := range srcInfos {
-			copySemaphore.Acquire(ctx, 1)
+			_ = copySemaphore.Acquire(ctx, 1)
```
[the remark this answers is not in the document]
Comment (vrothberg) [reply copy/copy.go]: Agreed, we need to check the error as we shouldn't continue when we couldn't acquire the semaphore.

File: docker/daemon/daemon_dest.go [hunk 25]
```diff
@@ -73,7 +73,7 @@ func imageLoadGoroutine(ctx context.Context, c *client.Client, reader *io.PipeRe
 		if err == nil {
 			reader.Close()
 		} else {
-			reader.CloseWithError(err)
+			_ = reader.CloseWithError(err) //always returns nil
```
[the remark this answers is not in the document]
Comment (vrothberg) [reply docker/daemon/daemon_dest.go]: It is (see below). But I am okay if we decide to handle the errors to pro-actively avoid bad surprises if the behavior changes in the future.

```golang
// CloseWithError closes the writer; subsequent reads from the  
// read half of the pipe will return no bytes and the error err,
// or EOF if err is nil.                                        
//                                                              
// CloseWithError always returns nil.                           
func (w *PipeWriter) CloseWithError(err error) error {          
        return w.p.CloseWrite(err)                              
}                                                                                                       
````

File: docker/tarfile/dest.go [hunk 26]
```diff
@@ -113,7 +113,7 @@ func (d *Destination) PutBlob(ctx context.Context, stream io.Reader, inputInfo t
 		if err != nil {
 			return types.BlobInfo{}, err
 		}
-		_, err = streamCopy.Seek(0, os.SEEK_SET)
+		_, err = streamCopy.Seek(0, io.SeekCurrent) // os.SEEK_SET is deprecated: Use io.SeekStart, io.SeekCurrent, and io.SeekEnd.
```
[the remark this answers is not in the document]
Comment (vrothberg) [reply docker/tarfile/dest.go]: Right. I wanted this comment to be in the commit message. It's important to explain why certain changes were performed. Not only for reviewing but also when looking up the code (and history) in the future.

File: oci/archive/oci_dest.go [hunk 27]
```diff
@@ -43,7 +44,10 @@ func (d *ociArchiveImageDestination) Reference() types.ImageReference {
 // Close removes resources associated with an initialized ImageDestination, if any
 // Close deletes the temp directory of the oci-archive image
 func (d *ociArchiveImageDestination) Close() error {
-	defer d.tempDirRef.deleteTempDir()
+	defer func() {
+		err := d.tempDirRef.deleteTempDir()
+		logrus.Debugf("error deleting tmp dir: %v", err)
```
[the remark this answers is not in the document]
Comment (vrothberg) [reply oci/archive/oci_dest.go]: I have a weak preference for logging it. But as @mtrmac mentions above, we should only log when `err == nil`. Also logs should start with a capital letter to make it easier to read/brain parse.

File: signature/signature.go (hunk 10 above)
Comment (vrothberg) [reply signature/signature.go]: Sure, but why are we doing this change? Instead of ignoring the error, we're now assigning it to nowhere.

File: storage/storage_image.go [hunk 28]
```diff
@@ -148,6 +148,9 @@ func (s *storageImageSource) getBlobAndLayerID(info types.BlobInfo) (rc io.ReadC
 	// callers should try to retrieve layers using their uncompressed digests, so no need to
 	// check if they're using one of the compressed digests, which we can't reproduce anyway.
 	layers, err := s.imageRef.transport.store.LayersByUncompressedDigest(info.Digest)
+	if err != nil {
+		return nil, -1, "", err
+	}
```
[the remark this answers is not in the document]
Comment (vrothberg) [reply storage/storage_image.go]: Yes, we should not return here. This was not in the previous versions of the commits - please add a comment to GitHub when new changes are added.

We can ignore this error here.

File: storage/storage_test.go (hunk 16 above)
Comment (vrothberg) [reply storage/storage_test.go]: This belongs into the commit message not at the place were variables have been removed.

File: docker/daemon/daemon_dest.go (hunk 25 above)
[the remark this answers is not in the document]
Comment (mtrmac) [reply docker/daemon/daemon_dest.go]: That’s Pipe**Writer**; the same commitment does not exist for Pipe**Reader**.

File: docker/archive/transport_test.go [hunk 29]
```diff
@@ -76,13 +75,6 @@ func testParseReference(t *testing.T, fn func(string) (types.ImageReference, err
 	}
 }
 
-// refWithTagAndDigest is a reference.NamedTagged and reference.Canonical at the same time.
-type refWithTagAndDigest struct{ reference.Canonical }
-
-func (ref refWithTagAndDigest) Tag() string {
-	return "notLatest"
-}
```
[the remark this answers is not in the document]
Comment (vrothberg) [reply docker/archive/transport_test.go]: This is still open. @bege13mot please reply to the comments.

File: openshift/openshift-copies.go (hunk 23 above)
Comment (bege13mot) [reply openshift/openshift-copies.go]: Sorry, I'm not clear about these copy&pasted errors

Comment (mtrmac) [reply openshift/openshift-copies.go]: The text of the 3 `Debugf` messages is the same (“closing client CA”), but only one of them actually closes a client CA; the other two should be updated to match what actually failed (client certificate/key file).

File: docker/tarfile/dest.go [hunk 30]
```diff
@@ -113,7 +113,7 @@ func (d *Destination) PutBlob(ctx context.Context, stream io.Reader, inputInfo t
 		if err != nil {
 			return types.BlobInfo{}, err
 		}
-		_, err = streamCopy.Seek(0, os.SEEK_SET)
+		_, err = streamCopy.Seek(0, io.SeekCurrent)
```
[the remark this answers is not in the document]
Comment (bege13mot) [reply docker/tarfile/dest.go]: Complaint - Deprecated: Use io.SeekStart, io.SeekCurrent, and io.SeekEnd

File: docker/tarfile/src.go [hunk 31]
```diff
@@ -146,7 +146,7 @@ func (s *Source) openTarComponent(componentPath string) (io.ReadCloser, error) {
 	}
 	if header.FileInfo().Mode()&os.ModeType == os.ModeSymlink { // FIXME: untested
 		// We follow only one symlink; so no loops are possible.
-		if _, err := f.Seek(0, os.SEEK_SET); err != nil {
+		if _, err := f.Seek(0, io.SeekCurrent); err != nil {
```
[the remark this answers is not in the document]
Comment (bege13mot) [reply docker/tarfile/src.go]: Deprecated: Use io.SeekStart, io.SeekCurrent, and io.SeekEnd

**Comment by vrothberg** [comment]
@bege13mot, can you rebase? We had some CI flakes that are fixed now.

One thing I just noticed now is that all commits are not signed with your real name but with the GitHub user which is in conflict with the contributing policies (https://github.com/containers/image/blob/master/CONTRIBUTING.md). My apologies for noticing that late.

**Comment by mtrmac** [comment]
This is now very big, and likely enough to break if any other PR is merged. Maybe it would be more practical to split it into smaller PRs? (Definitely no need to split it _artificially_ just because I said so now — but, if the rebase happens one commit at a time, maybe separate 10 unrelated commits into a separate PR.)

Ideally, of course, a rebased version would pass tests, be quickly reviewed, and immediately merged; splitting this is distinctly second best. Still, a possible option.

**Comment by vrothberg** [comment]
I think something went wrong while rebasing since there are many older commits unrelated to this PR.

**Comment by vrothberg** [comment]
Failures look legit:
```
package github.com/golangci/golangci-lint: no Go files in /gopath/src/github.com/golangci/golangci-lint

Makefile:58: recipe for target 'tools.timestamp' failed
```

**Comment by rhatdan** [comment]
@vrothberg @mtrmac Tests pass, is this ready to merge?
@bege13mot thanks for all of the work on this?

**Comment by vrothberg** [comment]
> @vrothberg @mtrmac Tests pass, is this ready to merge?
> @bege13mot thanks for all of the work on this?

I'll do one final review now :+1:

File: storage/storage_test.go (hunk 15 above)
Comment (bege13mot) [reply storage/storage_test.go]: Unfortunately, I can't find the usage

File: Makefile [hunk 32]
```diff
@@ -56,6 +56,7 @@ tools: tools.timestamp
 
 tools.timestamp: Makefile
 	@GO111MODULE="off" go get -u $(BUILDFLAGS) golang.org/x/lint/golint
+	curl -sfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh| sudo sh -s -- -b $(go env GOPATH)/bin v1.21.0
```
[the remark this answers is not in the document]
Comment (bege13mot) [reply Makefile]: Unfortunately, it isn't help. 
`install: cannot create regular file '//golangci-lint': Permission denied``

**Comment by rhatdan** [comment]
LGTM

File: Makefile (hunk 32 above)
[the remark this answers is not in the document]
Comment (mtrmac) [reply Makefile]: ⚠️ 

“help” or not, this project _absolutely must not_ create files in the root directory of the user’s filesystem!

If using `sudo` here really has that effect, it must be removed, pretty much at any cost, up to and including not running golang-ci, if there were really no alternative.

File: storage/storage_test.go (hunk 15 above)
Comment (mtrmac) [reply storage/storage_test.go]: No “usage” is necessary. The lines are effective tests as they are (if `storage…` does not implement the interface, this will report an error).

The lines should be restored, and the warning, if any, silenced using a comment or whatever golang-CI uses.

File: Makefile (hunk 32 above)
[the remark this answers is not in the document]
Comment (vrothberg) [reply Makefile]: I'll fix that. $(go env GOPATH) didn't evaluate here for some odd reason.

File: storage/storage_test.go (hunk 15 above)
Comment (vrothberg) [reply storage/storage_test.go]: I can take care of that.

Comment (vrothberg) [reply storage/storage_test.go]: Done (will open a PR in a couple of mins)

Comment (mtrmac) [reply storage/storage_test.go]: Fixed in https://github.com/containers/image/pull/762 , for the record.
